keps/sig-node/20190806-memory-manager.md

```diff
@@ -0,0 +1,264 @@
+- [Proposal](#proposal)
+  - [Proposed Changes](#proposed-changes)
+    - [New Component: Memory Manager](#new-component-memory-manager)
+      - [Calculate NUMA node affinity for memory and hugepages](#calculate-numa-node-affinity-for-memory-and-hugepages)
+      - [Provide topology hint for Topology Manager](#provide-topology-hint-for-topology-manager)
+      - [Isolate memory and hugepages for a container](#isolate-memory-and-hugepages-for-a-container)
+      - [New Interfaces](#new-interfaces)
+    - [Topology Manager changes](#topology-manager-changes)
+    - [Internal Container Lifecycle changes](#internal-container-lifecycle-changes)
+    - [Feature Gate and Kubelet Flags](#feature-gate-and-kubelet-flags)
+- [Graduation Criteria](#graduation-criteria)
+  - [Phase 1: Alpha (target v1.1x)](#phase-1-Alpha-target-v11x)
+  - [Phase 2: Beta](#phase-1-Beta)
+  - [GA (stable)](#ga-stable)
+
+# Overview
+
+NUMA Awareness is well known for a solution to boost performance for diverse use cases including DPDK and Database. Kubernetes support NUMA related features for NUMA sensitive containers.
+The features(CPU Manager, Device Manager, Topology Manager) treat CPU, Device and it's NUMA topology, but there's no feature for memory and hugepages. It can cause inter-NUMA node communication for Memory access which leads to increase an I/O latency and decrease performance.
+
+The Memory Manager is proposed for a solution deploying Pod and Containers with Guaranteed QoS class of Memory and Hugepages under NUMA awareness.
+
+# Motivation
+
+## Related Features
+
+- [Node Topology Manager][node-topology-manager] is a feature that collects topology hint from various hint providers to calculate socket affinity for a container. Topology Manager judge container whether admit or not under configured policy and socket affinity.
+
+- [CPU Manager][cpu-manager] is a feature provides a solution for CPU Pinning based on cgroups cpuset subsystem, it also offer topology hint to Topology Manager.
+
+- [Device Manager][device-manager] is a one of features provides topology hint for Topology Manager. The main objectivity of Device Manager is allow vendors to adverise their resources to kubelet.
+
+- [Hugepages][hugepages] is a feature allows container to consume pre-allocated hugepages with pod isolation of huge pages.
+
+- [Node Allocatable Feature][node-allocatable-feature] is a feature helps to reserve compute resources to prevent resource starvation. The kube-reserved and system-reserved is used to reserve resources for kubelet and system(OS, etc). Now following resources are supported to reserve: cpu, memory, ephemral storage.
+
+## Related issues
+
+- [Hardware topology awareness at node level (including NUMA)][numa-issue]
+- [Support Container Isolation of Hugepages][hugepage-issue]
+- [Support isolating memory to single NUMA node][memory-issue]
+
+## Goals
+
+- Guarantee alignment of resources for isolating container's Memory and Hugepages with CPU and I/O devices.
```

Comment:
This topic has come up before, but I don't believe anyone has produced a comprehensive demonstration of the before / after effect of hugepages alignment. Do you have some test results on recent Linux kernel that shows 1) without explicit memory affinity and with single-socket CPU affinity, misaligned hugepages are accessed 2) performance penalty for a representative application (e.g. DPDK packet forwarding) when hugepages are misaligned (in the worst case).

cc @sjenning

Reply:
I know it is somewhat anecdotal because I couldn't share the exact test results last time the topic came up, but we have actually performed tests related to point 2) with our 4G, and 5G RAN CNFs.
the results were quite disappointing: 20-50ms latency spikes were observed for an application whose E2E latency figures should be <1ms-ish according to the standards. 
In case we need hard data to move this topic forward I can try to pursue this in the company, probably we could partially disclose the test setup, and results

Reply:
Curious about whether cpuset.mems is the best option here or whether the current cpuset.cpus plus the default local_allocation memory policy is already doing a good job in most cases.

I suppose the case you especially want to guard against is when the local numa node becomes exhausted and hugepages allocation falls back to a non-local node? Would setting your own memory policy to "bind" instead? The downside is that you won't know the app is doomed until later when you try to allocate memory.

In any case, it would be great to add some discussion of these alternatives along with pros/cons somewhere in this doc.

Reply:
see https://www.kernel.org/doc/Documentation/vm/numa_memory_policy.txt

Reply:
Regarding the question from @Connor, my team tested a misaligned NIC case but it was not a case you mentioned.
It seems that we can refer @Levovar 's test result.

As @Connor mentioned above, I especially want to guard against misaligned Hugepages case, I believe this is the worst case for DPDK users.
In this case, the DPDK app will cross NUMA at least more than 3 times.
1) polling packet from a buffer of NIC to Hugepages.
2) copying pointers of packets from RX/TX ring buffer on Hugepages to a conventional memory buffer.
3) getting access to real packet object through a pointer

I believe NUMA memory policy can't be a solution.
I have two reasons.
1) To use this memory policy, the process(in a container) should apply memory policy for itself by using set_mempolicy.
This function requires nodemask that represents NUMA node.
But the process doesn't know that know how many processes are running on a NUMA node and their resource affinity.
So, it is hard to determine NUMA node for bind policy by process itself.
2) The downside you mentioned is critical, there is no way to guarantee resources that are requested with Guaranteed QoS.
But we can schedule container to specific NUMA node by using cpuset subsystem and we can also guarantee Guaranteed QoS.

`it would be great to add some discussion of these alternatives along with pros/cons somewhere in this doc.`
=> I agree with it, in fact, my original plan was using NUMA memory policy for this feature.

Reply:
I am with @ConnorDoyle on this.  I would also like to see this misalignment happen naturally, if possible, and have the effect measured.

Modern linux kernels have automatic NUMA balancing and it should be that, for a process pinned to cpus on a single socket/node, the memory allocations are satisfied from the local memory node assuming sufficient capacity.  In this case, managing the CPU implicitly manages the memory/hugepages as well.

There is a conflating discussion here with the NIC being remote wrt to the DPDK workload, but that is the job of the device/topology managers to ensure this workload<->device locality right?

Reply:
how modern exactly are we talking about? because my experience is that this automatic balancing did not fully happen, so I will try and re-test it.
just curious which kernel version should behave like this. Is 4.18 good enough?

Reply:
It is in RHEL 7
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/virtualization_tuning_and_optimization_guide/sect-virtualization_tuning_optimization_guide-numa-auto_numa_balancing
```
echo 1 > /proc/sys/kernel/numa_balancing
```

Reply:
` I would also like to see this misalignment happen naturally`  
=> I think it can happen under below scenarios, it's a kind of resource exhausted case.

### Scenario A (Remote Hugepages Case)
- Container requests 2 exclusive cpus , 5GB Memory and 3x 1GB-Hugepages.
- deploy multiple sets of above container.

**[Idle State Case]**
![001](https://raw.githubusercontent.com/bg-chun/kep-draft/master/scenario-a-001.png)
When container is idle state, container may consume one or two hugepages.
In this case, there is no problem, there is no remote access of memory.
If containers always consume maximum size of hugepages, this case will not happen.  
(Kernel allocates hugepages when process allocates memory over `mmap`.
If process allocates less size of 1GB memory, kernel will allocate one hugepage.
If process allocates 1.5GB memory, kernel will allocate two hugepages to process)

**[Busy State Case]**
![002](https://raw.githubusercontent.com/bg-chun/kep-draft/master/scenario-a-002.png)
When container is busy state, container may consume maximum quantity of hugepages.  
Here, it's three of 1GB-Hugepages.
In this case, if there is no extra hugepages on local numa node, container will consume hugepages on another NUMA node.

### Scenario B (Remote Memory Case)
- Container Blue requests 2 exclusive cpus, 95GB Memory and 3x 1GB-Hugepages.
- Container Red requests 2 exclusive cpus , 5GB Memory and 3x 1GB-Hugepages.

**[Idle State Case]**
![003](https://raw.githubusercontent.com/bg-chun/kep-draft/master/scenario-a-003.png)  
In this scenario, container blue can be such kind of In-Memory Database(redis).  
Or it can be multiple containers that sharing cpus.  
When container red is idle state, there will be no remote access of memory.

**[Busy State Case]**
![004](https://raw.githubusercontent.com/bg-chun/kep-draft/master/scenario-a-004.png)  
But when container red is busy state, there will be remote access of memory.

![005](https://raw.githubusercontent.com/bg-chun/kep-draft/master/scenario-a-005.png)
Maybe, this worstest case can happen, this case will reduce performance of both containers.

## My conclusion
Maybe you guys would say that above scenarios are not normal case or it's just a corner case.
But I want to say that it can happen because cgroup sets only limits of memory regardless of memory capacity of NUMA node.

The point is `kubelet` does not consider memory capacity of NUMA node.
Topology Manager, CPU Manager and Device Manager (will) does it for cpu and devices(such as nic and gpu), but there is no consideration of Memory(including Hugepages) at all.

As @sjenning mentioned above, it seems that [numa_balancing](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/virtualization_tuning_and_optimization_guide/sect-virtualization_tuning_optimization_guide-numa-auto_numa_balancing) can cover most general cases(like idle state in scenarios).  
But it cannot **guarantee** local access of memory for container.  
Without `guaranteeing local access of memory`, we cannot guarantee performance of DPDK containers.  
To guarantee local access of memory, kubelet should schedule container to a NUMA node based on container's memory request.  
Then enforce container to consume memory on a NUMA node using a `cupset.mems`.  
Below image shows a concept of scheduling.  

![concept-of-scheduling](https://raw.githubusercontent.com/bg-chun/kep-draft/master/concept-of-scheduling.png)


Otherwise, Openstack which is used to run VNF(DPDK) fully guarantee it, through `hw:numa_mem`property, see [here](https://docs.openstack.org/nova/rocky/admin/cpu-topologies.html#customizing-instance-numa-placement-policies) and [here](https://docs.openstack.org/nova/rocky/user/flavors.html#extra-specs).  
Moreover Openstack also support that VM takes certain amount of memory from multiple NUMA nodes, see table 3.2 [here](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_openstack_platform/7/html/instances_and_images_guide/ch-manage_instances#section-update-flavor-metadata).  
It seems that Libvirt driver support it, but i'm not sure, i was just an one of users of Openstack.

Example
``` 
Example when the instance has 8 vCPUs and 4GB RAM:
    hw:numa_nodes=2
    hw:numa_cpus.0=0,1,2,3,4,5
    hw:numa_cpus.1=6,7
    hw:numa_mem.0=3  //Mapping 3 GB of RAM to NUMA node 0
    hw:numa_mem.1=1  //Mapping 1 GB of RAM to NUMA node 1. 
```

Reply:
Thanks for the detailed response.  I agree that there are cases where auto NUMA balancing doesn't work, mostly because we are already pinning the cpus.

Because the Linux scheduler does not have the freedom to move the tasks from one socket to another (might not be good for device locality and it might also not be able to migration 1GB huge pages; not sure), NUMA balancing is backed into a corner for the most part.

Said another way, once you start pinning a resource, you have to pin all the resources because you've reduced the freedom of the kernel to place tasks and migrate resources automatically.  Hence the proliferation of hint providing managers we have.

At the end of the day, we have already set about making a second-level scheduler in the kubelet for this use case and we will need a hint provider (scheduler predicate) for every resource that the primary kube scheduler considers (cpu, memory, hugepages, devices).

So I'm on board with the need for a memory manager in the kubelet.

Enforcing `cpuset.mems` to a single node might cause issues if a pod is scheduled to a (kubernetes) node that, among all (numa) nodes, has the available resources but no single (numa) node has the available resources.  Is the idea to deny admission in that case?  Or include both nodes in `cpuset.mems` and let remote accesses happen?

It seem like a heuristic on the cpu manager side to handle the majority of the scenarios mentioned is to allocate cpus from the least utilized (in terms of pinned cpus allocated) numa node.  That trivially handles the situation for the first n pods, where n is the number of numa nodes, as long as the workloads don't request more than any one node can provide.. and the nodes have equal memory capacity.  It would continue to work after that as long as the cpu requests are proportion to the memory/hugepages requests (not sure if that is normally the case).  The cpu manager might already do this afaik.

Reply:
_Enforcing cpuset.mems to a single node might cause issues if a pod is scheduled to a (kubernetes) node that, among all (numa) nodes, has the available resources but no single (numa) node has the available resources._
=>As I know correct, in [Topology Manager](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0035-20190130-topology-manager.md), the concept of `strict` policy is to reject a pod if alignment cannot be satisfied. Basically, the concept of this feature is the same.

_Is the idea to deny admission in that case? Or include both nodes in cpuset.mems and let remote accesses happen?_
=>Yes, the idea is to deny admission but that is not role of this feature, this feature will provide `nil` or negative topology hint(0, 0, false) to Topology Manager. I expect admission will be denied when a worker node is configured `strict` policy for Topology Manager.

_The cpu manager might already do this afaik._
=> The heuristic way you have mentioned it feels like a best-effort solution. As you said, it would continue to work, if enough hugepages and memory are available for pod admission on a numa node. I think it is far from guaranteeing local-access to memory which this feature will guarantee.